libraries/stdlib/test/ranges/RangeTest.kt

```diff
@@ -5,11 +5,198 @@
 
 package test.ranges
 
+import kotlin.math.min
+import kotlin.math.sign
 import kotlin.test.*
 
+fun <T : Comparable<T>, S> testProgressionCollection(
+    limitRange: Collection<T>,
+    stepRange: Iterable<S>,
+    makeRange: (T, T, S) -> Collection<T>,
+    nextItem: (T, S) -> T?,
+    sign: (S) -> Int
+) {
+    for (start in limitRange)
+        for (finish in limitRange)
+            for (step in stepRange) {
+                val s = sign(step).takeIf { it != 0 } ?: continue
+                val range = makeRange(start, finish, step)
+                val listRange = generateSequence(start) { nextItem(it, step) }
+                    .takeWhile { if (s > 0) start <= it && it <= finish else start >= it && it >= finish }.toList()
+                try {
+                    collectionWithIterating(listRange, range, limitRange)
+                    collectionWithoutIterating(listRange, object : Collection<T> {
```

Comment:
A lot of range and progression iteration cases are already covered in [`RangeIterationTest`](https://github.com/jetbrains/kotlin/blob/651d6c9d514e6ac620f1f18ff32728a6c050a573/libraries/stdlib/test/ranges/RangeIterationTest.kt). I don't think we should be duplicating iteration tests here considering that we didn't change the implementation of iterator.

Reply:
I have just looked there. It really repeats collectionWithIterating so it can be remobed but doesn't do the same with `without`. `Without` version checks size, contains and containsAll while `RangeIterationTestBase` doesn't. It also checks, that real iterator is not used. Furthermore, it is applied to big progressions which cannot be fully stored in memory.

I suppose that
```kotlin
if (expectedElements.isEmpty())
    assertTrue(sequence.none())
else
    assertEquals(expectedElements, sequence.toList())
```
should be replaced with
```kotlin
assertEquals(expectedElements.isEmpty(). sequence.none())
assertEquals(expectedElements, sequence.toList())
```

in [RangeIterationTest.kt](https://github.com/JetBrains/kotlin/blob/651d6c9d514e6ac620f1f18ff32728a6c050a573/libraries/stdlib/test/ranges/RangeIterationTest.kt#L77) as isEmpty is defined separately from iterator and progression implementation that always returns true would pass your test.

Reply:
Thanks for the suggestion, makes sense to apply it to `RangeIterationTestBase.doTest`

Reply:
Therefore, I'll remove `collectionWithIterating` tests and simplify the remaining ones.

Reply:
Great!